lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp

```diff
@@ -254,14 +265,50 @@ static Value createConstantOp(FIRRTLType opType, APInt value,
   return Value();
 }
 
-static Type getHandshakeBundleDataType(BundleType bundle) {
-  if (auto dataType = bundle.getElementType("data")) {
-    auto intType = dataType.cast<firrtl::IntType>();
-    return IntegerType::get(bundle.getContext(), intType.getWidthOrSentinel(),
+/// Creates a Value that has an assigned zero value. For bundles, this
+/// corresponds to assigning zero to each element recursively.
+static Value createZeroDataConst(FIRRTLType dataType, Location insertLoc,
+                                 OpBuilder &builder) {
+  return TypeSwitch<Type, Value>(dataType)
+      .Case<IntType>([&](auto dataType) {
+        return createConstantOp(dataType,
+                                APInt(dataType.getBitWidthOrSentinel(), 0),
+                                insertLoc, builder);
+      })
+      .Case<BundleType>([&](auto bundleType) {
+        auto wire = builder.create<WireOp>(insertLoc, bundleType);
+
+        for (size_t i = 0, e = bundleType.getNumElements(); i < e; ++i) {
+          auto field = builder.create<SubfieldOp>(insertLoc, wire, i);
+          Value zeroData =
+              createZeroDataConst(field.getType(), insertLoc, builder);
+          builder.create<ConnectOp>(insertLoc, field, zeroData);
+        }
+        return wire;
+      })
+      .Default([&](Type) -> Value { llvm_unreachable("Unknown type"); });
+}
+
```

Comment:
FYI, you can create zero constant bundles with bitcast from constant, such as: https://github.com/llvm/circt/blob/90ae0886aa06020d9f370118bb69dc1d3afbddcb/lib/Dialect/FIRRTL/Transforms/SFCCompat.cpp#L104-L109

Reply:
This does indeed simplify things substantially. Thanks for pointing this out to me.